File: Dumplings/src/main/java/net/sinclairstudios/android/dumplings/domain/DumplingServingCalculation.java
package net.sinclairstudios.android.dumplings.domain;

public final class DumplingServingCalculation {
    private final Dumpling dumpling;
    private Fraction servings;

    public DumplingServingCalculation(Dumpling dumpling, Fraction servings) {
        this.dumpling = dumpling;
        setServings(servings);
    }

    private void setServings(Fraction servings) {
        this.servings = servings;
    }

    public void transferRemainderTo(DumplingServingCalculation remainderRecipient) {
        Fraction remainder = servings.getRemainder();

        remainderRecipient.setServings(remainderRecipient.getServings().add(remainder));
        setServings(getServings().subtract(remainder));
    }

    public Fraction getServings() {
        return servings;
    }

    public Dumpling getDumpling() {
        return dumpling;
    }
}

File: app/src/main/java/net/sinclairstudios/dumplings/domain/DumplingServingCalculation.kt
package net.sinclairstudios.dumplings.domain
import java.io.Serializable

class DumplingServingCalculation(val dumpling : Dumpling, var servings: Fraction) : Serializable {

    fun trimRemainder() : Fraction {
        val remainder = this.servings.getRemainder()
        this.servings = this.servings.subtract(remainder)
        return remainder
    }

    fun addServings(servings: Fraction) {
        this.servings = this.servings.add(servings)
    }

    public override fun toString() : String {
        return "$dumpling with $servings servings"
    }
}